test/functional/tests/io_class/test_io_class_stats_file_size_core.py

```diff
@@ -0,0 +1,198 @@
+from api.cas.cache_config import CacheMode, CleaningPolicy, SeqCutOffPolicy
+from api.cas.ioclass_config import IoClass
+from core.test_run_utils import TestRun
+from storage_devices.disk import DiskTypeSet, DiskType, DiskTypeLowerThan
+from test_tools import fs_utils
+from test_tools.disk_utils import Filesystem
+from test_tools.fio.fio import Fio
+from test_tools.fio.fio_param import IoEngine, ReadWrite
+from test_utils.os_utils import sync, drop_caches
+from test_utils.size import Size, Unit
+
+
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_io_class_stats_file_size_core():
+    """
+    title: Open CAS statistics values for file size - per core.
+    description: Check Open CAS ability to assign correct IO class to file size.
+    pass_criteria:
+        - after FIO without direct statistics increases only for tested IO class
+    """
+
+    with TestRun.step("Prepare devices."):
+        cache_devices, core_devices = prepare_devices()
+
+    with TestRun.step("Start caches (one for each supported cache mode) and add core devices."):
+        caches, cores = prepare_caches_and_cores(cache_devices, core_devices)
+
+    with TestRun.step("Make filesystem on OpenCAS devices and mount it."):
+        for core, fs in zip(cores, cycle(Filesystem)):
+            mount_point = core.path.replace('/dev/', '/mnt/')
+            core.create_filesystem(fs)
+            core.mount(mount_point)
+            sync()
+
+    with TestRun.step("In default IO class configuration file set Eviction priority=1 "
+                      "and load it for all caches."):
+        io_classes = prepare_io_classes(caches)
+
+    with TestRun.step("Run fio for all devices checking IO class statistics."):
+        sizes = get_sizes(io_classes)
+        size_min = Size(512, Unit.Byte)
+
+        for io_class, size in zip(io_classes[2:-1], sizes):
+            with TestRun.step(f"Testing IO class {io_class.id}."):
+                with TestRun.step("Run fio for each device and check that "
+                                  "only for tested IO class statistics increases."):
+                    for core in cores:
+                        core.reset_counters()
+                        fio = fio_params(core, size_min, size)
+                        result = fio.run()
+                        sync()
+                        drop_caches()
+                        issued_reqs_no = \
+                            result[0].write_requests_number() + result[0].read_requests_number()
+                        served_reqs_no = \
+                            core.get_io_class_statistics(io_class.id).request_stats.requests_total
+                        check_statistics(core, io_class, issued_reqs_no, served_reqs_no)
+                        fs_utils.remove(f"{core.mount_point}/*", force=True, recursive=True)
+                    size_min = size + Unit.Byte
```

Comment:
If I understand this correctly the min_size doesn't conribute much to the test and results in surprising behavior:

in the first iteration fio is sent to two files: 512B and 4K. as a result all I/O belong to the same I/O class. On other hand in later iterations the smaller file is 512+n (inserting into ioclass no 11) and the larger file inserts to one of the other I/O classes. As a result it's hard to tell exactly what is the expected occupancy after each iteration.
in all iterations except the first one the job writing to a small file has the following specification: file_size=512+n, block_size=512, time based. So effectively a single block is being hit all the time by fio. Adding extra byte to the min_size only results in a strangely unaligned file_size and does not trigger any interesting behavior.
I suggest to:

remove the min_file logic, just have a single fio job in each iteration hitting file sized accordingly to current io_class.
modify check_statistics() so that it verifies that the number of I/O reported by fio and IO class statistics are exactly equal rather than <=.

Reply:
First fio:
io_class 11 size_min=512 size_max=4096
Second fio:
io_class 12 size_min=4097 size_max=16384
Third fio:
io_class 13 size_min=16385 size_max=65536 
etc.
In each iteration size_min is from start of io_class and size_max is from the end of io_class.

Reply:
I see, thanks